File: app/src/main/java/com/a6v/tjreader/activities/MainActivity.java
package com.a6v.tjreader.activities;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.NavigationView;
import android.support.v4.app.Fragment;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;

import com.a6v.tjreader.R;
import com.a6v.tjreader.app.Logger;
import com.a6v.tjreader.fragments.FeedFragment;
import com.a6v.tjreader.net.FeedSorting;
import com.a6v.tjreader.presenters.MainMenuPresenter;
import com.a6v.tjreader.services.ArticleService;
import com.a6v.tjreader.services.DataSourceManager;
import com.a6v.tjreader.services.FeedType;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import dagger.Provides;
import dagger.Subcomponent;

public class MainActivity extends BaseActivity
  implements NavigationView.OnNavigationItemSelectedListener, ActivityComponentProvider
{
  private static final FeedSorting DEFAULT_SORTING = FeedSorting.RECENT;

  @BindView(R.id.toolbar)
  Toolbar toolbar;

  private AbstractActivityComponent component;
  private MainMenuPresenter menuPresenter;

  @Inject DataSourceManager dataSourceManager;

  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    component = getAppComponent().mainActivityComponent(new Module(this,
      getLastCustomNonConfigurationInstance()));
    setContentView(R.layout.activity_feed);
    ButterKnife.bind(this);
    toolbar.inflateMenu(R.menu.feed);
    menuPresenter = new MainMenuPresenter();
    menuPresenter.attachView(findViewById(R.id.drawer), this);
    toolbar.setOnMenuItemClickListener(item -> {
      menuPresenter.open();
      return true;
    });

    if (savedInstanceState == null) {
      menuPresenter.setChecked(R.id.navi_feed_all);
      showFeedFragment(FeedType.ALL, DEFAULT_SORTING);
    }
  }

  @Override
  public void onBackPressed() {
    if (!menuPresenter.close()) {
      super.onBackPressed();
    }
  }

  @Override
  public Object onRetainCustomNonConfigurationInstance() {
    return dataSourceManager;
  }

  @Override
  public AbstractActivityComponent getActivityComponent() {
    return component;
  }

  @Override
  public boolean onNavigationItemSelected(@NonNull MenuItem item) {
    menuPresenter.close();
    switch (item.getItemId()) {
      case R.id.navi_feed_all:
        showFeedFragment(FeedType.ALL, DEFAULT_SORTING);
        break;
      case R.id.navi_feed_news:
        showFeedFragment(FeedType.NEWS, DEFAULT_SORTING);
        break;
      case R.id.navi_feed_articles:
        showFeedFragment(FeedType.ARTICLES, DEFAULT_SORTING);
        break;
      case R.id.navi_feed_offtop:
        showFeedFragment(FeedType.OFF_TOPIC, DEFAULT_SORTING);
        break;
      case R.id.navi_saved:
        break;
      case R.id.navi_read:
        break;
    }
    return true;
  }

  private void showFragment(Fragment fragment) {
    getSupportFragmentManager()
      .beginTransaction()
      .replace(R.id.placeholder, fragment)
      .commit();
  }

  private void showFeedFragment(FeedType type, FeedSorting sorting) {
    showFragment(FeedFragment.create(type, sorting));
  }

  @Subcomponent(modules = Module.class)
  @PerActivity
  public interface Component extends AbstractActivityComponent {
    void inject(FeedFragment fragment);
  }

  @dagger.Module
  public static class Module extends ActivityModule {
    private final DataSourceManager dataSourceManager;

    public Module(Activity activity, Object lastCustomNonConfigurationInstance) {
      super(activity);
      if (lastCustomNonConfigurationInstance != null) {
        dataSourceManager = (DataSourceManager) lastCustomNonConfigurationInstance;
      } else {
        dataSourceManager = null;
      }
    }

    @Provides
    @PerActivity
    public DataSourceManager provideDataSourceManager(ArticleService articleService,
      Logger logger)
    {
      if (dataSourceManager == null) {
        return new DataSourceManager(articleService, logger);
      }
      return dataSourceManager;
    }
  }
}

File: app/src/main/kotlin/com/a6v/tjreader/activities/MainActivity.kt
package com.a6v.tjreader.activities

import android.content.res.Configuration
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.view.GravityCompat
import android.support.v7.app.ActionBarDrawerToggle
import android.view.Menu
import android.view.MenuItem
import com.a6v.tjreader.R
import com.a6v.tjreader.extensions.getAppComponent
import com.a6v.tjreader.fragments.FeedFragment
import com.a6v.tjreader.fragments.ReadFragment
import com.a6v.tjreader.fragments.SavedFragment
import com.a6v.tjreader.service.FeedSorting
import com.a6v.tjreader.service.FeedType
import com.a6v.tjreader.widgets.NavigationDrawer
import dagger.Subcomponent
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : BaseActivity() {
  lateinit var component: MainActivityComponent

  override fun initComponent() {
    component = getAppComponent().mainActivity(ActivityModule(this))
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)
    setSupportActionBar(toolbar)
    component.inject(this)
    NavigationDrawer(drawer_menu, drawer, enabled = true) {
      when (it.itemId) {
        R.id.navi_feed_all -> showFeedFragment(FeedType.ALL, FeedSorting.RECENT)
        R.id.navi_feed_news -> showFeedFragment(FeedType.NEWS, FeedSorting.RECENT)
        R.id.navi_feed_articles -> showFeedFragment(FeedType.ARTICLES, FeedSorting.RECENT)
        R.id.navi_feed_offtop -> showFeedFragment(FeedType.OFF_TOPIC, FeedSorting.RECENT)
        R.id.navi_saved -> showFragment(SavedFragment())
        R.id.navi_read -> showFragment(ReadFragment())
      }
    }
    if (savedInstanceState == null) {
      showFeedFragment(FeedType.ALL, FeedSorting.RECENT)
      drawer_menu.setCheckedItem(R.id.navi_feed_all)
    }
  }

  override fun onCreateOptionsMenu(menu: Menu?): Boolean {
    menuInflater.inflate(R.menu.main, menu)
    return true
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    when (item.itemId) {
      R.id.main_menu_drawer -> {
        //todo add toggle drawer
        if (drawer.isDrawerOpen(GravityCompat.END)) {
          drawer.closeDrawers()
        } else {
          drawer.openDrawer(GravityCompat.END)
        }
        return true
      }
    }
    return super.onOptionsItemSelected(item)
  }

  override fun onBackPressed() {
    if (drawer.isDrawerOpen(GravityCompat.END)) {
      drawer.closeDrawers()
    } else {
      super.onBackPressed()
    }
  }

  fun showFragment(fragment: Fragment) {
    supportFragmentManager.beginTransaction().replace(R.id.placeholder, fragment).commit()
  }

  fun showFeedFragment(feedType: FeedType, sorting: FeedSorting) {
    showFragment(FeedFragment.create(feedType, sorting))
  }
}

@Subcomponent(modules = arrayOf(ActivityModule::class))
@PerActivity
interface MainActivityComponent : AbstractActivityComponent {
  fun inject(activity: MainActivity)

  fun inject(fragment: FeedFragment)

  fun inject(fragment: SavedFragment)

  fun inject(fragment: ReadFragment)
}